ultratrace2/model/files/loaders/dicom.py

```diff
@@ -0,0 +1,79 @@
+from PIL import Image  # type: ignore
+
+import numpy as np
+import os
+import pydicom  # type: ignore
+
+from .base import FileLoadError, ImageSetFileLoader
+
+
+class DICOMLoader(ImageSetFileLoader):
+    def __init__(self, path: str, pixels: np.ndarray):
+        """Construct the DICOMLoader from a pixel array.
+
+        The `shape` of the pixel array should be `(n_frames, n_rows, n_columns)` for greyscale
+        images and `(n_frames, n_rows, n_columns, rgb_data)` for full-color images."""
+        super().__init__(path)
+        self.pixels = pixels
+        # FIXME: these should be in the `.ultratrace/` dir
+        self.png_dir = f"{self.path}-frames"
+        if not os.path.exists(self.png_dir):
+            os.mkdir(self.png_dir, mode=0o755)
+
+    def is_greyscale(self) -> bool:
+        return len(self.pixels) == 3
+
+    def __len__(self) -> int:
+        return self.pixels.shape[0]
+
+    def get_png_filepath_for_frame(self, i: int) -> str:
+        return os.path.join(self.png_dir, f"{i:06}.png")
+
+    def get_frame(self, i: int) -> Image.Image:
+        png_filepath = self.get_png_filepath_for_frame(i)
+        if os.path.exists(png_filepath):
+            return Image.open(png_filepath)
+        else:
+            arr = (
+                self.pixels[i, :, :] if self.is_greyscale() else self.pixels[i, :, :, :]
+            )
+            img = Image.fromarray(arr)
+            img.save(png_filepath, format="PNG", compress_level=1)
+            return img
+
+    @classmethod
+    def from_file(cls, path: str) -> "DICOMLoader":
+
+        if not os.path.exists(path):
+            raise FileLoadError("Cannot load from path: '{path}': cannot find")
+
+        try:
+            dicom = pydicom.read_file(path)
+        except pydicom.errors.InvalidDicomError as e:
+            raise FileLoadError(str(e))
```

Comment:
You might want to use exception chaining so that you don't erase the stack trace from the caught exception (which may be useful for debugging), e.g.
```suggestion
            raise FileLoadError("my error message") from e
```
see the sixth paragraph of the introduction in https://docs.python.org/3/library/exceptions.html

Reply:
Cool, didn't know about this feature. Done in 336f414